File: Python/04search_insert_position.py
nums=input("nums[]=")
target=int(input("target="))
nums=nums.split(" ")
nums=[eval(i) for i in nums]
nums.sort()
left=0
right=len(nums)-1

#当target比列表中所有元素都大时
if nums[len(nums)-1] < target:
        nums.append(target)
        print(len(nums)-1)
        print(nums)
else:
        while left < right:#二分法查找
                mid=(left+right)//2
                if nums[mid] < target:
                        left=mid+1
                        nums.insert(left,target)
                        print(left)
                        print(nums)
                elif nums[mid] == target:
                        nums.insert(mid,target)
                        print(mid)
                        print(nums)
                elif nums[mid] == target:
                        right=mid-1
                        nums.insert(left+1,target)
                        print(left+1)
                        print(nums)
                elif nums[left] == target:
                        nums.insert(left,target)
                        print(left)
                        print(nums)
                        break
                elif nums[right] == target:
                        nums.insert(right,target)
                        print(right)
                        print(nums)
                        break
        
        
        

File: c/04_search_insert_position.c
//author: Qingqing Guo, sunnyguoqq@outlook

#include<stdio.h>
#define MAXSIZE 100

/*Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.  */

int main()
{
	int nums[MAXSIZE];
	int len, target;
	int temp;
	int left = 0;
	int right = 0;
	int mid = 0;

	/*input data*/
	printf("len=");
	scanf_s("%d", &len);
	printf("target=");
	scanf_s("%d", &target);
	printf("nums=");
	for (int i = 0; i <= len; i++)
	{
		scanf_s("%d", &nums[i]);
	}

	/*sorting*/
	for (int i = 0; i <= len; i++)
	{
		for (int j = 0; j < len - i; j++)
		{
			if (nums[j] > nums[j + 1])
			{
				temp = nums[j];
				nums[j] = nums[j + 1];
				nums[j + 1] = temp;
			}
		}
	}
	/*find and insert position*/
	if (nums[len] < target)
	{
		printf("%d", len + 1);
	}
	else
		right = len;
		while (left < right)
		{
			mid = left + (right - left) / 2;
			if (nums[mid] == target)
			{
				printf("%d", mid);
				break;
			}
			else if (nums[mid] < target)
			{
				left = mid + 1;
				printf("%d", left);
			}
			else if (nums[mid] > target&&target>nums[mid-1])
			{
				printf("%d", mid);
				break;
			}
			else if (nums[left] >= target)
			{
				printf("%d", left);
				break;
			}
		}
	return 0;
}

